analyzer/src/main/kotlin/Analyzer.kt

```diff
@@ -40,6 +40,7 @@ import org.ossreviewtoolkit.model.VcsInfo
 import org.ossreviewtoolkit.model.config.AnalyzerConfiguration
 import org.ossreviewtoolkit.model.config.RepositoryConfiguration
 import org.ossreviewtoolkit.model.readValue
+import org.ossreviewtoolkit.spdx.VCS_DIRECTORIES
```

Comment:
I believe doing this commit `as-is` is wrong. Let's consider an example:

`
project/package.json
project/src/include/x.h
project/src/include/x.cpp
`
If the above is a native project (where we do not support any package manager) applying this commit would eliminate the unmnaged project. All files would be associated with the `package.json`. Then you want to exclude the `package.json`, because the project is native only. If you do so, you effectively exclude also all native code as it gets only associated with the `package-json`.

Reply:
While it's correct that before this PR an unmanaged project would have been created, the `project/src/include/x.h` and `project/src/include/x.cpp` files from your example would not be associated to it, but to the NPM project defined by `project/package.json`. In general, every file is associated to the "closest" project (defined by a definition file) when walking along the parents in directory tree.

So when you say

> All files would be associated with the `package.json`.

then that's correct, but unless I'm terribly mistaken, it's exactly how it was before.

Reply:
> would not be associated to it, but to the NPM project defined by `project/package.json`.

Are you sure this is true? I believe the files are assigned to both projects, `unmanaged` and `project/package.json`.

Reply:
Actually all files are associated to all definition files found in their parent directories. For example if you have:

```
- README.md
- a
  - build.gradle
  - b
    - package.json
    - c
      - main.cpp
```

Then `main.cpp` would be associated with the Unmanager, Gradle and NPM projects, because we never filter subdirectories that contain other projects. I have suggested we do that at least for Unmanaged in the past, because the idea of this fake package manager was to make sure that also files that cannot be associated to a project get scanned, but it was never implemented.

@fviernau The scenario you describe above would be a weird project setup, but if the package.json was in the root directory you would have exactly the same issue already with the existing code. A workaround could be to not exclude the project but only all dependency scopes, and I wonder if that would not be sufficient, because Unmanaged was never intended to solve the problem you describe.

Reply:
>  The scenario you describe above would be a weird project setup

because?

Reply:
Let me try to re-phrase my problem with this change:

The idea to associate files with definitions files and to exclude files if their associated definition file is excluded has a problem.  If you have a project which does (1) not use a package manager or which does (2) use a package manager NOT supported by ort, you could not exclude any definition file which is unrelated to your project, but in your code base.
Because the exclusion would lead to the exclusion of files which should not be excluded.

This unmanaged implementation helped a lot with this problem. So however it was originally intended does not really matter, I believe we need to find a solution for the above problem before making this commit.

Reply:
> > The scenario you describe above would be a weird project setup
> 
> because?

Because you shouldn't put `*.h` / `*.cpp` files which are C++ files in the directory tree of an (unrelated) NPM project where only `*.js` files are expected. Or is this some funky-special NPM project that implements a [native addon](https://medium.com/jspoint/a-simple-guide-to-load-c-c-code-into-node-js-javascript-applications-3fcccf54fd32) and the `*.h` / `*.cpp` files *are* in fact part of the NPM project?

Reply:
> Because you shouldn't put *.h / *.cpp files which are C++ files in the directory tree of an (unrelated) NPM project where only *.js files are expected. Or is this some funky-special NPM project that implements a native addon and the *.h / *.cpp files are in fact part of the NPM project?

Agree that one shouldn't do that setup (probably) but should ORT really say that we don't support such setups ?

Reply:
> Agree that one shouldn't do that setup (probably) but should ORT really say that it doesn't support such setups ?

Actually, I indeed would be fine with ORT not supporting such setups. Also in other areas (like maintaining metdata) we leverage ORT to foster best engineering practices. So why not here?

Anyway, that was also not your original question; your original question was why it's weird. And you seem to agree it's weird now.

Reply:
> > The scenario you describe above would be a weird project setup
> 
> because?

Because why do you have a package.json lying around in your native code project?

Reply:
> > would not be associated to it, but to the NPM project defined by `project/package.json`.
> 
> Are you sure this is true? I believe the files are assigned to both projects, `unmanaged` and `project/package.json`.

For reference, turns out that @fviernau was right, and scan findings are not only associated to the "closest" project, but to all projects on the way up to the root... also see https://github.com/oss-review-toolkit/ort/issues/5365.